### DISCUSSION TITLE
mini.comment: change the default commentstring for a filetype

### DISCUSSION THREAD

**Discussion by krishna-bala** [nvim-mini/mini.nvim#332] [Q&A] [answered]
I would like to change the default commentstring for `c/cpp` files from `/* %s */` to `// %s` when using `mini.comment`. In the [mini-comment.txt](https://github.com/echasnovski/mini.comment/blob/main/doc/mini-comment.txt) documentation, it mentions that 

> Comment structure is inferred from 'commentstring': either from current
  buffer or from locally active tree-sitter language (only on Neovim>=0.9).

I am using neovim 0.9.0, so I believe the `commentstring` is inferred from the active tree-sitter language. Is there a way to modify the active tree-sitter language so that I can change the default `commentstring` for a given filetype?

**Comment by echasnovski** [comment, marked as answer]
> I would like to change the default commentstring for `c/cpp` files from `/* %s */` to `// %s` when using `mini.comment`. In the [mini-comment.txt](https://github.com/echasnovski/mini.comment/blob/main/doc/mini-comment.txt) documentation, it mentions that
> 
> > Comment structure is inferred from 'commentstring': either from current
> > buffer or from locally active tree-sitter language (only on Neovim>=0.9).
> 
> I am using neovim 0.9.0, so I believe the `commentstring` is inferred from the active tree-sitter language. Is there a way to modify the active tree-sitter language so that I can change the default `commentstring` for a given filetype?

Sure, by using one of the "usual" ways to modify 'commentstring' for a filetype:
- (highly suggested) Create an 'after/ftplugin/c.lua' file in your config directory (`~/.config/nvim` in Unix) and put `vim.bo.commentstring = '// %s'` there.
- Create autocommand for `FileType` event: `au FileType c setlocal commentstring=//\ %s`.

**Reply by krishna-bala** [reply]
Thank you, this was extremely helpful! Also thank you for the [reddit post on after/ftplugin](https://www.reddit.com/r/neovim/comments/x3zp6t/usage_of_afterftplugin_directory_for/), that was also very helpful. The first suggestion solves the problem for me. However, my (modified) attempt at the second version didn't work -- I'm describing it below in case anyone is interested, but 

## feel free to ignore since the first suggestion works.

Here's how I tried to create an autocmmand for `FileType` event:
I'm using LazyVim, and therefore have my `~/.config/nvim` structured per the [LazyVim documentation](https://www.lazyvim.org/configuration/general) like so:
```
|- ~/.config/nvim
   |- after/
      |- ftplugin/
         |- c.lua  (assume this is an empty file for this example)
   |- lua/
      |- config/
         |- autocmds.lua
         |- keymaps.lua
         |- lazy.lua
         |- options.lua
      |- plugins/
         |- ...
```

For configuring autocmds, the documentation says
> The files autocmds.lua, keymaps.lua, lazy.lua and options.lua under lua/config will be automatically loaded at the appropriate time, so you don't need to require those files manually. LazyVim comes with a set of default config files that will be loaded before your own.
> 
> DANGER
> Do not require autocmds, keymaps, lazy or options under lua/config/ or lazyvim.config manually. LazyVim will load those files automatically.

Therefore, I configured my `autocmds.lua` as follows:
```lua
-- Autocmds are automatically loaded on the VeryLazy event
-- Default autocmds that are always set: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/autocmds.lua
-- Add any additional autocmds here

local function augroup(name)
  return vim.api.nvim_create_augroup("lazyvim_config_" .. name, { clear = true })
end

vim.api.nvim_create_autocmd("FileType", {
  group = augroup("commentstring"),
  pattern = { "*.cc", "*.c", "*.h" },
  callback = function()
    vim.opt.commentstring = "// %s"
  end,
  desc = "Change commentstring for c/c++ files",
})
```

However, this does not seem to update the `commentstring` for `*.cc`, `*.c`, or `*.h` files.

**Reply by echasnovski** [reply]
> However, this does not seem to update the `commentstring` for `*.cc`, `*.c`, or `*.h` files.

One problem here is that `FileType` takes actual filetype as pattern (see `:h FileType`). You can try `pattern = { 'c', 'cpp' }`.

**Reply by mangkoran** [reply]
I stumbled across this issue too and found this solution helpful. Don't mind me I just want to put the [documentation](https://neovim.io/doc/user/autocmd.html#FileType) regarding the "pattern" behaviour of "FileType" events 😅

Thank you for the solution!
